lib/js-api/src/externals/memory.rs

```diff
@@ -0,0 +1,319 @@
+use crate::export::{Export, VMMemory};
+use crate::exports::{ExportError, Exportable};
+use crate::externals::Extern;
+use crate::store::Store;
+use crate::{MemoryType, MemoryView};
+use std::convert::TryInto;
+use thiserror::Error;
+
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
+use wasmer_types::{Bytes, Pages, ValueType};
+
+/// Error type describing things that can go wrong when operating on Wasm Memories.
+#[derive(Error, Debug, Clone, PartialEq, Hash)]
+pub enum MemoryError {
+    /// The operation would cause the size of the memory to exceed the maximum or would cause
+    /// an overflow leading to unindexable memory.
+    #[error("The memory could not grow: current size {} pages, requested increase: {} pages", current.0, attempted_delta.0)]
+    CouldNotGrow {
+        /// The current size in pages.
+        current: Pages,
+        /// The attempted amount to grow by in pages.
+        attempted_delta: Pages,
+    },
+    /// A user defined error value, used for error cases not listed above.
+    #[error("A user-defined error occurred: {0}")]
+    Generic(String),
+}
+
+#[wasm_bindgen]
+extern "C" {
+    /// [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Memory)
+    #[wasm_bindgen(js_namespace = WebAssembly, extends = js_sys::Object, typescript_type = "WebAssembly.Memory")]
+    #[derive(Clone, Debug, PartialEq, Eq)]
+    pub type JSMemory;
+
+    /// The `grow()` protoype method of the `Memory` object increases the
+    /// size of the memory instance by a specified number of WebAssembly
+    /// pages.
+    ///
+    /// Takes the number of pages to grow (64KiB in size) and returns the
+    /// previous size of memory, in pages.
+    ///
+    /// [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Memory/grow)
+    #[wasm_bindgen(catch, method, js_namespace = WebAssembly)]
+    pub fn grow(this: &JSMemory, pages: u32) -> Result<u32, JsValue>;
+}
```

Comment:
Already implemented, https://github.com/rustwasm/wasm-bindgen/blob/b97837b0989b58595f3ef72bd5691ce420591204/crates/js-sys/src/lib.rs#L3674-L3683. It's not marked as `catch` though. Is it the difference here? Is it on-purpose? If yes, then it must be documented here, and maybe open a PR to `wasm-bindgen`?

Reply:
I've seen this https://github.com/rustwasm/wasm-bindgen/pull/2599, so it's fine. It must be written in the documentation. I'm adding a suggestion above.

Reply:
Yeah, that's the difference.